File: 2398. Maximum Number of Robots Within Budget/Solution2398.cpp
#include<iostream>
#include<vector>
#include<deque>

using namespace std;
class Solution2398 {
public:
    int maximumRobots(vector<int>& chargeTimes, vector<int>& runningCosts, long long budget) {
        int left = 0,  n = chargeTimes.size();
        long cursum = 0L;
        deque<int> d;
        for (int right = 0; right < n; right++) {
            cursum += runningCosts[right];
            while (!d.empty() && chargeTimes[d.back()] <= chargeTimes[right]) {
                d.pop_back();
            }
            d.push_back(right);
            if (chargeTimes[d.front()] + (right - left + 1) * cursum > budget) {
                if (d.front() == left) {
                    d.pop_front();
                }
                cursum -= runningCosts[left];
                left++;
            }
        }
        return n - left;
    }
};

File: 2398. Maximum Number of Robots Within Budget/Solution2398.py
class Solution2398:
    def maximumRobots(self, chargeTimes: List[int], runningCosts: List[int], budget: int) -> int:
        left, right,cursum, n = 0,0,0, len(chargeTimes)
        d = deque()
        for right in range(n):
            cursum += runningCosts[right]
            while d and chargeTimes[d[-1]] <= chargeTimes[right]:
                d.pop()
            d.append(right)
            if chargeTimes[d[0]]+(right - left+1)*cursum > budget:
                if d[0] == left:
                    d.popleft()
                cursum -= runningCosts[left] 
                left += 1
        return n - left